src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.ICompilationTracker.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal partial class SolutionState
+    {
+        private interface ICompilationTracker
+        {
+            /// <summary>
+            /// Returns true if this tracker currently either points to a compilation, has an in-progress
+            /// compilation being computed, or has a skeleton reference.  Note: this is simply a weak
+            /// statement about the tracker at this exact moment in time.  Immediately after this returns
+            /// the tracker might change and may no longer have a final compilation (for example, if the
+            /// retainer let go of it) or might not have an in-progress compilation (for example, if the
+            /// background compiler finished with it).
+            /// 
+            /// Because of the above limitations, this should only be used by clients as a weak form of
+            /// information about the tracker.  For example, a client may see that a tracker has no
+            /// compilation and may choose to throw it away knowing that it could be reconstructed at a
+            /// later point if necessary.
+            /// </summary>
+            bool HasCompilation { get; }
+            ProjectState ProjectState { get; }
+
+            ICompilationTracker Clone();
```

Comment:
what'st he value in cloning?  since these are immutable already, right?

Reply:
This is used to implement https://github.com/dotnet/roslyn/blob/02428e32a5e34dbaef1c27be3851ae8256a43841/src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs#L1548-L1562

Reply:
let's remove in a followup pr.